<?xml version="1.0"?>
<!--
Software License Agreement (BSD)

\file      stereo.launch
\authors   Michael Hosmar <mhosmar@clearpathrobotics.com>
\copyright Copyright (c) 2018, Clearpath Robotics, Inc., All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that
the following conditions are met:
 * Redistributions of source code must retain the above copyright notice, this list of conditions and the
   following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
   following disclaimer in the documentation and/or other materials provided with the distribution.
 * Neither the name of Clearpath Robotics nor the names of its contributors may be used to endorse or promote
   products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WAR-
RANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, IN-
DIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->
<launch>
  <arg name="record" default="1" />

  <!-- Common parameters -->
  <arg name="camera_name"               default="kitchen" />
  <arg name="device_type"               default="GigE" />

  <!-- When unspecified, the driver will use the default framerate as given by the
      camera itself. Use the parameter 'control_frame_rate' to enable manual frame
      rate control, and 'frame_rate' to set the frame rate value. -->
  <arg name="control_frame_rate"        default="true" />
  <arg name="frame_rate"                default="22" />

  <arg name="isp_enable"                default="False" />
  <arg name="encoding"                  default="BayerRG8" />
  <arg name="color_balance"             default="Continuous" /> <!-- Off, Once, or Continuous -->

  <arg name="PG1_camera_serial"        default="11292568" />
  <arg name="PG1_camera_calibrated"    default="0" />

  <arg name="PG2_camera_serial"       default="11292573" />
  <arg name="PG2_camera_calibrated"   default="0" />

  <arg name="PG3_camera_serial"       default="11292571" />
  <arg name="PG3_camera_calibrated"   default="0" />

  <arg name="PG4_camera_serial"       default="11292592" />
  <arg name="PG4_camera_calibrated"   default="0" />

  <arg name="PG5_camera_serial"       default="11292590" />
  <arg name="PG5_camera_calibrated"   default="0" />

  <arg name="PG6_camera_serial"       default="11210206" />
  <arg name="PG6_camera_calibrated"   default="0" />

  <!-- NODELET manager -->
  <!-- Both cameras are not loaded into one nodelet manager to avoid the namespacing issue. -->
  <node pkg="nodelet" type="nodelet" name="camera_nodelet_manager" args="manager" />

  <group ns="$(arg camera_name)" >
    <group ns="PG1" >
      <!-- Camera nodelet -->
      <node pkg="nodelet" type="nodelet" name="camera_nodelet"
            args="load spinnaker_camera_driver/SpinnakerCameraNodelet /camera_nodelet_manager" >

        <param name="frame_id"                        value="camera_PG1" />
        <param name="serial"                          value="$(arg PG1_camera_serial)" />
        <param name="device_type"                     value="$(arg device_type)" />

        <!-- Frame rate -->
        <param name="acquisition_frame_rate_enable"   value="$(arg control_frame_rate)" />
        <param name="acquisition_frame_rate"          value="$(arg frame_rate)" />

        <param name="isp_enable"                      value="$(arg isp_enable)" />
        <param name="auto_white_balance"              value="$(arg color_balance)" />
        <param name="image_format_color_coding"       value="$(arg encoding)" />

        <param name="exposure_auto"  value="Off"/>
        <param name="exposure_time"  value="40000"/>

        <param name="auto_gain" value="Off"/>
        <param name="gain" value="6.0"/>

        <!-- Use the camera_calibration package to create this file -->
        <param name="camera_info_url" if="$(arg PG1_camera_calibrated)"
               value="file://$(env HOME)/.ros/camera_info/$(arg PG1_camera_serial).yaml" />
      </node>

      <!-- Debayering nodelet -->
      <node pkg="nodelet" type="nodelet" name="image_proc_debayer"
          args="load image_proc/debayer /camera_nodelet_manager">
      </node>

      <!-- Image view Node -->
      <node name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
        <remap from="image" to="/kitchen/PG1/image_raw"/>
        <param name="autosize" value="true" />
      </node>

      <group if="$(arg record)">
        <!-- record image data -->
        <node
            pkg="rosbag"
            type="record"
            name="rosbag_record"
            args="record -o /media/chaoliu/Elements/$(arg camera_name)-PG1 -b 0 /$(arg camera_name)/PG1/image_raw"/>
      </group>

      <!-- save recorded video file -->
      <!-- <node name="extract_images" pkg="image_view" type="video_recorder" respawn="false" output="screen"> -->
      <!--   <remap from="image" to="/kitchen/left/image_raw"/> -->
      <!--   <param name="filename" value="/home/chaoliu/Desktop/left.avi"/> -->
      <!--   <param name="fps" value="22"/> -->
      <!-- </node> -->
    </group>

    <group ns="PG2" >
      <!-- Camera nodelet -->
      <node pkg="nodelet" type="nodelet" name="camera_nodelet"
            args="load spinnaker_camera_driver/SpinnakerCameraNodelet /camera_nodelet_manager" >

        <param name="frame_id"                        value="camera_PG2" />
        <param name="serial"                          value="$(arg PG2_camera_serial)" />
        <param name="device_type"                     value="$(arg device_type)" />

        <!-- Frame rate -->
        <param name="acquisition_frame_rate_enable"   value="$(arg control_frame_rate)" />
        <param name="acquisition_frame_rate"          value="$(arg frame_rate)" />

        <param name="isp_enable"                      value="$(arg isp_enable)" />
        <param name="auto_white_balance"              value="$(arg color_balance)" />
        <param name="image_format_color_coding"       value="$(arg encoding)" />

        <param name="exposure_auto"  value="Off"/>
        <param name="exposure_time"  value="35000"/>

        <param name="auto_gain" value="Off"/>
        <param name="gain" value="5.0"/>

        <!-- Use the camera_calibration package to create this file -->
        <param name="camera_info_url" if="$(arg PG2_camera_calibrated)"
               value="file://$(env HOME)/.ros/camera_info/$(arg PG2_camera_serial).yaml" />
      </node>

      <!-- Debayering nodelet -->
      <node pkg="nodelet" type="nodelet" name="image_proc_debayer"
          args="load image_proc/debayer /camera_nodelet_manager">
      </node>

      <!-- Image view Node -->
      <node name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
        <remap from="image" to="/kitchen/PG2/image_raw"/>
        <param name="autosize" value="true" />
      </node>

      <group if="$(arg record)">
        <!-- record image data -->
        <node
            pkg="rosbag"
            type="record"
            name="rosbag_record"
            args="record -o /media/chaoliu/Elements/$(arg camera_name)-PG2 -b 0 /$(arg camera_name)/PG2/image_raw"/>
      </group>

      <!-- save recorded video file -->
      <!-- <node name="extract_images" pkg="image_view" type="video_recorder" respawn="false" output="screen"> -->
      <!--   <remap from="image" to="/kitchen/right/image_raw"/> -->
      <!--   <param name="filename" value="/home/chaoliu/Desktop/right.avi"/> -->
      <!--   <param name="fps" value="22"/> -->
      <!-- </node> -->
    </group>

    <group ns="PG3" >
      <!-- Camera nodelet -->
      <node pkg="nodelet" type="nodelet" name="camera_nodelet"
            args="load spinnaker_camera_driver/SpinnakerCameraNodelet /camera_nodelet_manager" >

        <param name="frame_id"                        value="camera_PG3" />
        <param name="serial"                          value="$(arg PG3_camera_serial)" />
        <param name="device_type"                     value="$(arg device_type)" />

        <!-- Frame rate -->
        <param name="acquisition_frame_rate_enable"   value="$(arg control_frame_rate)" />
        <param name="acquisition_frame_rate"          value="$(arg frame_rate)" />

        <param name="isp_enable"                      value="$(arg isp_enable)" />
        <param name="auto_white_balance"              value="$(arg color_balance)" />
        <param name="image_format_color_coding"       value="$(arg encoding)" />

        <param name="exposure_auto"  value="Off"/>
        <param name="exposure_time"  value="25000"/>

        <param name="auto_gain" value="Off"/>
        <param name="gain" value="2.0"/>

        <!-- Use the camera_calibration package to create this file -->
        <param name="camera_info_url" if="$(arg PG3_camera_calibrated)"
               value="file://$(env HOME)/.ros/camera_info/$(arg PG3_camera_serial).yaml" />
      </node>

      <!-- Debayering nodelet -->
      <node pkg="nodelet" type="nodelet" name="image_proc_debayer"
          args="load image_proc/debayer /camera_nodelet_manager">
      </node>

      <!-- Image view Node -->
      <node name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
        <remap from="image" to="/kitchen/PG3/image_raw"/>
        <param name="autosize" value="true" />
      </node>

      <group if="$(arg record)">
        <!-- record image data -->
        <node
            pkg="rosbag"
            type="record"
            name="rosbag_record"
            args="record -o /media/chaoliu/ACTIONNET1/$(arg camera_name)-PG3 -b 0 /$(arg camera_name)/PG3/image_raw"/>
      </group>

      <!-- save recorded video file -->
      <!-- <node name="extract_images" pkg="image_view" type="video_recorder" respawn="false" output="screen"> -->
      <!--   <remap from="image" to="/kitchen/right/image_raw"/> -->
      <!--   <param name="filename" value="/home/chaoliu/Desktop/right.avi"/> -->
      <!--   <param name="fps" value="22"/> -->
      <!-- </node> -->
    </group>

    <group ns="PG4" >
      <!-- Camera nodelet -->
      <node pkg="nodelet" type="nodelet" name="camera_nodelet"
            args="load spinnaker_camera_driver/SpinnakerCameraNodelet /camera_nodelet_manager" >

        <param name="frame_id"                        value="camera_PG4" />
        <param name="serial"                          value="$(arg PG4_camera_serial)" />
        <param name="device_type"                     value="$(arg device_type)" />

        <!-- Frame rate -->
        <param name="acquisition_frame_rate_enable"   value="$(arg control_frame_rate)" />
        <param name="acquisition_frame_rate"          value="$(arg frame_rate)" />

        <param name="isp_enable"                      value="$(arg isp_enable)" />
        <param name="auto_white_balance"              value="$(arg color_balance)" />
        <param name="image_format_color_coding"       value="$(arg encoding)" />

        <param name="exposure_auto"  value="Off"/>
        <param name="exposure_time"  value="35000"/>

        <param name="auto_gain" value="Off"/>
        <param name="gain" value="5.0"/>

        <!-- Use the camera_calibration package to create this file -->
        <param name="camera_info_url" if="$(arg PG4_camera_calibrated)"
               value="file://$(env HOME)/.ros/camera_info/$(arg PG4_camera_serial).yaml" />
      </node>

      <!-- Debayering nodelet -->
      <node pkg="nodelet" type="nodelet" name="image_proc_debayer"
          args="load image_proc/debayer /camera_nodelet_manager">
      </node>

      <!-- Image view Node -->
      <node name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
        <remap from="image" to="/kitchen/PG4/image_raw"/>
        <param name="autosize" value="true" />
      </node>

      <group if="$(arg record)">
        <!-- record image data -->
        <node
            pkg="rosbag"
            type="record"
            name="rosbag_record"
            args="record -o /media/chaoliu/ACTIONNET1/$(arg camera_name)-PG4 -b 0 /$(arg camera_name)/PG4/image_raw"/>
      </group>

      <!-- save recorded video file -->
      <!-- <node name="extract_images" pkg="image_view" type="video_recorder" respawn="false" output="screen"> -->
      <!--   <remap from="image" to="/kitchen/right/image_raw"/> -->
      <!--   <param name="filename" value="/home/chaoliu/Desktop/right.avi"/> -->
      <!--   <param name="fps" value="22"/> -->
      <!-- </node> -->
    </group>

    <group ns="PG5" >
      <!-- Camera nodelet -->
      <node pkg="nodelet" type="nodelet" name="camera_nodelet"
            args="load spinnaker_camera_driver/SpinnakerCameraNodelet /camera_nodelet_manager" >

        <param name="frame_id"                        value="camera_PG5" />
        <param name="serial"                          value="$(arg PG5_camera_serial)" />
        <param name="device_type"                     value="$(arg device_type)" />

        <!-- Frame rate -->
        <param name="acquisition_frame_rate_enable"   value="$(arg control_frame_rate)" />
        <param name="acquisition_frame_rate"          value="$(arg frame_rate)" />

        <param name="isp_enable"                      value="$(arg isp_enable)" />
        <param name="auto_white_balance"              value="$(arg color_balance)" />
        <param name="image_format_color_coding"       value="$(arg encoding)" />

        <param name="exposure_auto"  value="Off"/>
        <param name="exposure_time"  value="35000"/>

        <param name="auto_gain" value="Off"/>
        <param name="gain" value="5.0"/>

        <!-- Use the camera_calibration package to create this file -->
        <param name="camera_info_url" if="$(arg PG5_camera_calibrated)"
               value="file://$(env HOME)/.ros/camera_info/$(arg PG5_camera_serial).yaml" />
      </node>

      <!-- Debayering nodelet -->
      <node pkg="nodelet" type="nodelet" name="image_proc_debayer"
          args="load image_proc/debayer /camera_nodelet_manager">
      </node>

      <!-- Image view Node -->
      <node name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
        <remap from="image" to="/kitchen/PG5/image_raw"/>
        <param name="autosize" value="true" />
      </node>

      <group if="$(arg record)">
        <!-- record image data -->
        <node
            pkg="rosbag"
            type="record"
            name="rosbag_record"
            args="record -o /media/chaoliu/ACTIONNET2/$(arg camera_name)-PG5 -b 0 /$(arg camera_name)/PG5/image_raw"/>
      </group>

      <!-- save recorded video file -->
      <!-- <node name="extract_images" pkg="image_view" type="video_recorder" respawn="false" output="screen"> -->
      <!--   <remap from="image" to="/kitchen/right/image_raw"/> -->
      <!--   <param name="filename" value="/home/chaoliu/Desktop/right.avi"/> -->
      <!--   <param name="fps" value="22"/> -->
      <!-- </node> -->
    </group>

    <group ns="PG6" >
      <!-- Camera nodelet -->
      <node pkg="nodelet" type="nodelet" name="camera_nodelet"
            args="load spinnaker_camera_driver/SpinnakerCameraNodelet /camera_nodelet_manager" >

        <param name="frame_id"                        value="camera_PG6" />
        <param name="serial"                          value="$(arg PG6_camera_serial)" />
        <param name="device_type"                     value="$(arg device_type)" />

        <!-- Frame rate -->
        <param name="acquisition_frame_rate_enable"   value="$(arg control_frame_rate)" />
        <param name="acquisition_frame_rate"          value="$(arg frame_rate)" />

        <param name="isp_enable"                      value="$(arg isp_enable)" />
        <param name="auto_white_balance"              value="$(arg color_balance)" />
        <param name="image_format_color_coding"       value="$(arg encoding)" />

        <param name="exposure_auto"  value="Off"/>
        <param name="exposure_time"  value="35000"/>

        <param name="auto_gain" value="Off"/>
        <param name="gain" value="5.0"/>

        <!-- Use the camera_calibration package to create this file -->
        <param name="camera_info_url" if="$(arg PG6_camera_calibrated)"
               value="file://$(env HOME)/.ros/camera_info/$(arg PG6_camera_serial).yaml" />
      </node>

      <!-- Debayering nodelet -->
      <node pkg="nodelet" type="nodelet" name="image_proc_debayer"
          args="load image_proc/debayer /camera_nodelet_manager">
      </node>

      <!-- Image view Node -->
      <node name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
        <remap from="image" to="/kitchen/PG6/image_raw"/>
        <param name="autosize" value="true" />
      </node>

      <group if="$(arg record)">
        <!-- record image data -->
        <node
            pkg="rosbag"
            type="record"
            name="rosbag_record"
            args="record -o /media/chaoliu/ACTIONNET2/$(arg camera_name)-PG6 -b 0 /$(arg camera_name)/PG6/image_raw"/>
      </group>

      <!-- save recorded video file -->
      <!-- <node name="extract_images" pkg="image_view" type="video_recorder" respawn="false" output="screen"> -->
      <!--   <remap from="image" to="/kitchen/right/image_raw"/> -->
      <!--   <param name="filename" value="/home/chaoliu/Desktop/right.avi"/> -->
      <!--   <param name="fps" value="22"/> -->
      <!-- </node> -->
    </group>
  </group>
</launch>
